### ISSUE TITLE
High difference between Decode TER and Streaming TER

### ISSUE THREAD

**Issue by MohitTare** [facebookresearch/wav2letter#834]
We have trained a streaming convnet hindi model using recipe provided [here](https://github.com/facebookresearch/wav2letter/tree/master/recipes/streaming_convnets). However after converting it using streaming tds converter we are getting large difference in TER as compared to the decoder result even after using same hyperparams. As per the paper ideally the difference should be less than 4%. Any idea on why it is so and anything we can do to bridge the high TER gap

**Comment by vineelpratap** [comment]
Hi,
 Could you give more details on the type of mistakes you are seeing and how you are measuring the TER for non streaming and streaming ASR. Is the trained acoustic model same for both systems ?

**Comment by MohitTare** [comment]
Ok here are the specifics

The error for non streaming version is measured by using Decoder binary provided. For the streaming ASR we are using the [MulithreadedStreamingBinary](https://github.com/facebookresearch/wav2letter/blob/v0.2/inference/inference/examples/MultithreadedStreamingASRExample.cpp)  and error rate is measured using Lavenshtein distance using below code snippet

```
import Levenshtein as Lev
def cer(s1, s2):
    """
    Arguments:
        s1 (string): target
        s2 (string): prediction
    Returns:
        (Errors, Target length)
    """
    s1 = ' '.join(s1.strip().split())
    s3 = ' '.join(s2.strip().split())
    length = len(s1)
    return Lev.distance(s1, s2), length

#to calculate mean cer
cers = []
for i in range(len(targets)):
    cers.append(cer(target[i],streaming_preds[i]))
cers = np.array(cers)
cer_mean = np.sum(cers[:, 0]) / np.sum(cers[:, 1])
```
The same hyperparams are used in streaming and decode binary version. The trained acoustic model is same for both systems and converted using the StreamingTDSConverter tool. 

Couple of observations on type of error
1. Repetition of words when using the streaming binary (which is not present in the output of decode binary)
2. Skipping of some words in the streaming binary and not in decode binary

Some observations we made on the above point when we tried using different training strategies
1. We observed that the difference is more (>19) when we use transfer learning (freeze all layers and just train last layers) for the language like hindi using english librispeech checkpoint 
2. When we train from scratch the error difference reduces (in range of 5-9)

Is it expected that we get same result using same hyperparams for Decoder binary which takes the entire audio file and the streaming binary (chunksize of 500msec) ?
Can transfer learning cause an issue ?

If needed i can also provide the config files we are using for training. Do let me know your thoughts on the same.

**Comment by vineelpratap** [comment]
> Couple of observations on type of error

While processing audio chunks, the current online decoder clears the decoder history till the frame corresponding to the last produced word. But, it suboptimal to do this as we are clearing the history after finding local optimum. 

I can look into improving this some time next week. In the mean time, could you try increasing the value here -  https://github.com/facebookresearch/wav2letter/blob/v0.2/inference/inference/examples/AudioToWords.cpp#L47 to 1000 or 1500. It should help a bit.  

> Can transfer learning cause an issue ?

For fine tuning, it might be better to allow ALL layers to be trained but possibly keeping a shorter schedule compared to full training. Also, how many word pieces do you use for hindi and what is the size of your dataset. Could you share the token set used for Hindi. For Hindi, it might be also worth to train the model on letters instead of word pieces. For this, you should consider changing the model to have have stride 4 or stride 2 instead of stride 8 present in the recipe.

**Comment by MohitTare** [comment]
Thanks for your input.

> https://github.com/facebookresearch/wav2letter/blob/v0.2/inference/inference/examples/AudioToWords.cpp#L47 to 1000 or 1500.

So we tried this, but the error rate didn't improved drastically (2-3 point diff).

I am sharing our token [set.](https://drive.google.com/file/d/17rlJdntqoAfo-lk9raeY7e7zQ_TmtZC2/view?usp=sharing) We have around 3000-4000 hrs of domain specific + generic hindi dataset and no of word pieces are 10000. Also we are training for ulaw encoded audios. Below are some of the other params in config file
```
--input=wav
--samplerate=8000
--criterion=ctc
--batchsize=60
--lr=0.46
--momentum=0.0
--maxgradnorm=0.5
--nthread=12
--mfsc=true
--usewordpiece=true
--wordseparator=_
--filterbanks=80
--minisz=200
--maxisz=18000
--mintsz=1
--sampletarget=0.02
--enable_distributed=true
--pcttraineval=1
--minloglevel=0
--logtostderr=1
--onorm=target
--sqnorm
--localnrmlleftctx=300
--use_saug=true
--saug_fmaskf=27
--saug_fmaskn=2
--saug_tmaskt=100
--saug_tmaskp=1.0
--saug_tmaskn=2
--iter=9999600
--stepsize=2499900
--saug_start_update=4999
--reportiters=1666
--gamma=1.0
```



We will try with letter based model instead of wordpiece based and let you know the result. Meanwhile any additional inputs will be really useful.

**Comment by MohitTare** [comment]
Ok so got the issue. The high difference was due to sampling frequency of 16k default [here](https://github.com/facebookresearch/wav2letter/blob/v0.2/inference/inference/module/feature/LogMelFeature.h#L31) 
Passing appropriate value of 8k for mulaw encoding fixed the issue and the TER is now within 4-5%. Still thanks for the suggestion for letter based model. We will try that.

**Comment by MohitTare** [comment]
Hey @vineelpratap we have made the changes now with stride 4 by changing [here](https://github.com/facebookresearch/wav2letter/blob/master/recipes/streaming_convnets/librispeech/am_500ms_future_context.arch#L19) 
Does this affect the future context of 500ms during streaming ?
We are getting a performance uplift during decoding but again facing high error in streaming.

We have now 1024 tokens (max token size = 2 word piece based ) and we are training with asg loss. Can this be an issue ?
